batchflow/models/torch/benchmark.py

```diff
@@ -0,0 +1,164 @@
+""" Time and memory trackers for benchmark operations with nn.Module. """
+import numpy as np
+import torch
+from ptflops import get_model_complexity_info
+
+from .utils import make_initialization_inputs
+
+# Different units for memory representation
+MEMORY_UNIT_CONSTANTS = {
+    'GB': 1 / (1024 ** 3),
+    'MB': 1 / (1024 ** 2),
+    'KB': 1 / 1024,
+    'B': 1.
+}
+
+class TimeTracker:
+    """ Measure time taken for an operation on GPU. """
+    def __init__(self):
+        self.start = None
+        self.end = None
+
+    def __enter__(self):
+        self.start = torch.cuda.Event(enable_timing=True)
+        self.end = torch.cuda.Event(enable_timing=True)
+        self.start.record()
+        return self
+
+    def __exit__(self, type, value, traceback):
+        self.end.record()
+        torch.cuda.synchronize()
+
+    @property
+    def value(self):
+        """ Get an operation time. """
+        return self.start.elapsed_time(self.end)
+
+class MemoryTracker:
+    """ Measure peak used memory for an operation on GPU. """
+    def __init__(self, device=None):
+        self.device = device
+        self.start_memory = None
+        self.end_memory = None
+
+    def __enter__(self):
+        torch.cuda.reset_peak_memory_stats()
+        self.start_memory = torch.cuda.max_memory_allocated(self.device)
+        return self
+
+    def __exit__(self, type, value, traceback):
+        self.end_memory = torch.cuda.max_memory_allocated(self.device)
+
+    @property
+    def value(self):
+        """ Get allocated memory for a module. """
+        return self.end_memory - self.start_memory
+
+def get_module_performance(module, inputs, n_repeats=300, warmup=40, device=None, track_backward=True,
+                           channels_last=False, amp=False, memory_unit='MB'):
+    """ Track module #macs, #parameters, time and memory consumption on forward and backward
+    pass for a given input tensor or inputs shape.
+
+    Parameters
+    ----------
+    module : nn.Module
+        Input module for which we track performance.
+    inputs : Tensor or sequence of ints
+        If Tensor, then it is a data which we use for tracking performance.
+        If sequence of ints, then it is a shape of tensor which will be generated for tracking performance.
+    n_repeats : int
+        Number of times to repeat forward and backward pass for tracking performance.
+    warmup : int
+        Number of starting iterations that won't be tracked.
+    device : str or torch.cuda.Device
+        Device for computations.
+        If str, then any option of device configuration from :class:`torch.nn.Module` is supported.
```

Comment:
`Device for computations. Any option of `PyTorch` device configuration is supported.`